oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/RegisteredClientJwtAssertionDecoderFactory.java

```diff
@@ -0,0 +1,157 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.server.authorization.authentication;
+
+import org.springframework.security.oauth2.core.ClientAuthenticationMethod;
+import org.springframework.security.oauth2.core.endpoint.OAuth2ParameterNames;
+import org.springframework.security.oauth2.jose.jws.JwsAlgorithms;
+import org.springframework.security.oauth2.jose.jws.MacAlgorithm;
+import org.springframework.security.oauth2.jose.jws.SignatureAlgorithm;
+import org.springframework.security.oauth2.jwt.JwtDecoder;
+import org.springframework.security.oauth2.jwt.JwtDecoderFactory;
+import org.springframework.security.oauth2.jwt.NimbusJwtDecoder;
+import org.springframework.security.oauth2.server.authorization.client.RegisteredClient;
+import org.springframework.security.oauth2.server.authorization.config.ClientSettings;
+import org.springframework.security.oauth2.server.authorization.config.ProviderSettings;
+import org.springframework.util.Assert;
+
+import javax.crypto.spec.SecretKeySpec;
+import java.nio.charset.StandardCharsets;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Creates JWT decoders for registered clients
+ *
+ * @author Rafal Lewczuk
+ * @since 0.2.1
+ */
+final class RegisteredClientJwtAssertionDecoderFactory implements JwtDecoderFactory<RegisteredClient> {
+
+	private static final Map<String, String> JCA_ALGORITHM_MAPPINGS;
+
+	static {
+		Map<String, String> mappings = new HashMap<>();
+		mappings.put(JwsAlgorithms.HS256, "HmacSHA256");
+		mappings.put(JwsAlgorithms.HS384, "HmacSHA384");
+		mappings.put(JwsAlgorithms.HS512, "HmacSHA512");
+		JCA_ALGORITHM_MAPPINGS = Collections.unmodifiableMap(mappings);
+	}
+
+	private final Map<String, CachedJwtDecoder> cachedDecoders = new ConcurrentHashMap<>();
+	private final ProviderSettings providerSettings;
+
+	RegisteredClientJwtAssertionDecoderFactory(ProviderSettings providerSettings) {
+		Assert.notNull(providerSettings, "providerSettings cannot be null");
+		this.providerSettings = providerSettings;
+	}
+
+	@Override
+	public JwtDecoder createDecoder(RegisteredClient registeredClient) {
+		Assert.notNull(registeredClient, "registeredClient cannot be null");
+
+		String signatureAlgorithm = registeredClient.getClientSettings().getTokenEndpointSignatureAlgorithm();
+		if (signatureAlgorithm == null) {
+			// default algorithm is RS256
+			signatureAlgorithm = JwsAlgorithms.RS256;
+		}
+		boolean isHmac = JCA_ALGORITHM_MAPPINGS.containsKey(signatureAlgorithm);
+
+		// No fancy .computerIfAbsent() calls as we need to check if client configuration has changed
```

Comment:
I've implement `buildDecoder()` to be (almost) identical as in `OidcIdTokenDecoderFactory` (error messages differ), but now it has less guards against  client misconfiguration. Missing checks are done when client registers, so it applies only to admin/operator misconfiguration.

It still checks if registered client configuration has changed but it's now simplified to this: `cachedDecoder != null && registeredClient.equals(cachedDecoder.registeredClient)`. I'd rather stick to it if possible, otherwise it means maintenance problems for those who configure static clients and periodically rotate keys.

Overall we have around 25 LOC less (plus validator class removed) - not much but from logic-dense regions.